Update concat service to run without loading WordPress

brandonpayton opened the pull request:
To increase performance, we are updating our concat service to run without loading WordPress. For this to work, we need to provide real FS paths to the concat script rather than WordPress URIs that need WordPress constants to properly resolve.

By default, we will continue to resolve WordPress resource URIs in the concat service, but we will work from real FS paths when the following constants are defined.

## PAGE_OPTIMIZE_CONCAT_BASE_DIR

The deepest common ancestor of all resources' filesystem paths. When defined, this will be used as the base directory for paths to concatenated resources.

For our sites, this will be `define( 'PAGE_OPTIMIZE_CONCAT_BASE_DIR', '/srv/htdocs' )`.

## PAGE_OPTIMIZE_ABSPATH

Some dependencies have paths based on ABSPATH, and page-optimize needs to know ABSPATH to determine their absolute filesystem path. But on our sites, ABSPATH points to a completely separate directory branch than the plugin and content directories.
ABSPATH = `/wordpress/core/5.3.2`
WP_CONTENT_DIR = `/srv/htdocs/wp-content`
WP_PLUGIN_DIR = `/srv/htdocs/wp-content/plugins`

This interferes with our plan to `define( 'PAGE_OPTIMIZE_CONCAT_BASE_DIR', '/srv/htdocs' )` because the ABSPATH-based paths resolve outside of `/srv/htdocs` under `/wordpress/core/5.3.2`.

But we have some hope because we symlink `/wordpress/core/5.3.2` as `/srv/htdocs/__wp__`, a path under `/srv/htdocs`. This enables us to add a workaround. If we take the dependencies based on `/wordpress/core/5.3.2` (ABSPATH) and base them on `/srv/htdocs/__wp__` instead, all our resource paths will be based on `/srv/htdocs` and will resolve properly when we attempt to concatenate them.

We can do this using `define( 'PAGE_OPTIMIZE_ABSPATH', '/srv/htdocs/__wp__' )`.

brandonpayton committed:
Support independent concat script

brandonpayton committed:
Use different term: DIR instead of PATH

brandonpayton committed:
Allow specifying different site root dir

brandonpayton committed:
Stop resolving symlinks

Resolving symlinks interferes with our ability to define
PAGE_OPTIMIZE_CONCAT_BASE_DIR as anything other than `/`.

brandonpayton committed:
Simplify name of plugin-specific ABSPATH constant

brandonpayton committed:
Clarify function name

brandonpayton committed:
Avoid file which uses WordPress functions

aidvu commented on page-optimize.php:
```diff
@@ -179,6 +183,46 @@ function page_optimize_sanitize_exclude_field( $value ) {
+	}
+
+	if ( empty( $new_path ) ) {
+		$new_path = '/page-optimize-resource-outside-base-path/' . basename( $original_fs_path );
```
Not sure why this?

Edit: I mean, if this is `page_optimize_use_concat_base_dir()` it will always return the weird path.

aidvu committed:
bump version, and stop doing anything in wp-admin

aidvu committed:
revert admin check

aidvu committed:
we have to treat both paths the same way

otherwise the CSS stuff that replaces relative
urls with absolute ones doesn't work

aidvu commented on page-optimize.php:
```diff
@@ -210,17 +210,26 @@ function page_optimize_use_concat_base_dir() {
+		return substr( $original_fs_path, strlen( $prefix_basedir ) );
+	}
+
+	return '/page-optimize-resource-outside-base-path/' . basename( $original_fs_path );
```
Still dislike. If we end up here, we shouldn't concat at all.

aidvu replied:
Can't think of anything better tho.

aidvu replied:
If we end up here, something shouldn't be concat in the first place.

brandonpayton replied:
yeah, I don't like it either. It was just my attempt to leave a clue rather than having everything just fail.

aidvu commented on service.php:
```diff
@@ -301,7 +301,13 @@ function page_optimize_get_path( $uri ) {
 
 	if ( defined( 'PAGE_OPTIMIZE_CONCAT_BASE_DIR' ) ) {
-		$path = realpath( PAGE_OPTIMIZE_CONCAT_BASE_DIR . "/$uri" );
+		if ( file_exists( PAGE_OPTIMIZE_CONCAT_BASE_DIR . "/$uri" ) ) {
```
Need to check both paths.

If you need reproducing, refresh `wp-admin` in the plugin editor, and you'll see it's trying to load the `spinner` from `__wp__/wp-includes`.

aidvu committed:
its ok, it only caches the headers sent below

aidvu commented on service.php:
```diff
@@ -66,10 +64,10 @@ function page_optimize_service_request() {
 	}
 
 	$output = page_optimize_build_output();
-	// TODO: It seems insecure to indiscriminately cache and send headers. Cache only specific headers.
```
Only caches stuff sent from the page generation below.

brandonpayton replied:
I think that is true in most cases, but it is technically possible for another plugin to run first and add headers. In WordPress, I'm not comfortable assuming we're the only one and in total control.

aidvu replied:
I'd return an array from `page_optimize_build_output` with `content` and `headers` so we can cache.

If you don't get to it today, will do tomorrow. :bowing_man:

brandonpayton replied:
Good idea! Now that you've said it, it feels like common sense :)

aidvu committed:
comments
